src/main/kotlin/com/mdrobnak/lalrpop/inspections/WrongInferredTypeInspection.kt

```diff
@@ -74,9 +72,24 @@ object WrongInferredTypeInspection : LocalInspectionTool() {
         }
     }
 
-    fun sameTypes(project: Project, lalrpopFile: PsiFile, importCode: String, type1: String, type2: String): Boolean {
+    fun sameTypes(
+        project: Project,
+        lalrpopFile: PsiFile,
+        importCode: String,
+        unitStructsGenerics: String,
+        type1: String,
+        type2: String
+    ): Boolean {
         val file = PsiFileFactory.getInstance(project)
-            .createFileFromText(RsLanguage, "mod __intellij_lalrpop {\n$importCode\ntype T1 = $type1;\ntype T2 = $type2;\n}")
+            .createFileFromText(
+                RsLanguage,
+                "mod __intellij_lalrpop {\n" +
+                        "    $importCode\n" +
+                        "    $unitStructsGenerics\n" +
+                        "    type T1 = $type1;\n" +
+                        "    type T2 = $type2;\n" +
+                        "}"
```

Comment:
This looks good, but I'm not sure if it's working (the only error shown here is from the Rust plugin):
![image](https://user-images.githubusercontent.com/4417660/103368655-24ed1880-4a96-11eb-8849-fbcd15559168.png)

Did you by chance try adding generic params to the type aliases (ex. `type T1<Rule1> = ...`) instead of unit structs? If alias generics work, it might be better for when the generic type has constraints on the generics, such that it would prevent unit structs from being used (ex. require `Debug`).

Reply:
> This looks good, but I'm not sure if it's working (the only error shown here is from the Rust plugin):
image

It is working as intended(as long as you don't try to apply quick fixes, which is a whole other rabbit hole I don't want to go down yet).

> Did you by chance try adding generic params to the type aliases (ex. `type T1<Rule1> = ...`) instead of unit structs? If alias generics work, it might be better for when the generic type has constraints on the generics, such that it would prevent unit structs from being used (ex. require `Debug`).

This could actually work *most* of the time, but I think you won't get very far without a way to get the rust `Substitution` object from somewhere before comparing the types, and the only way I think this would be possible is if you keep track of which macro arguments get passed down to this nonterminal, and run the inspection logic for each of them.

And even then, what will you do about the type parameters of `grammar<...>;`?

Constraints make this much harder than it needs to be. I wish lalrpop had a way to specify trait constraints on macro parameters, this would really help; without them it's a lot more like C++ templates rather than rust generics: unknown types coming in and the IDE just has to deal with it.

Needless to say, this really requires a lot more looking into, and is far outside the scope of this PR.